source/agora/node/Ledger.d

```diff
@@ -579,17 +656,77 @@ public class Ledger
                 this.params.CommonsBudgetAddress);
+            pkey_to_output_ind[this.params.CommonsBudgetAddress.data] = pkey_to_output_ind.length;
+        }
 
         // pay the validator for the past blocks
         if (auto payouts = this.fee_man.getAccumulatedFees(next_height))
             foreach (pair; payouts.byKeyValue())
                 if (pair.value > Amount(0))
+                {
                     coinbase_tx.outputs ~= Output(pair.value, pair.key);
+                    pkey_to_output_ind[pair.key.data] = pkey_to_output_ind.length;
+                }
+
+        ////////
+        // calculating block reward
+        ///////
+        if (block_reward_calc.is_payout_time(height))
+        {
+            immutable period_end = Height(height - params.BlockRewardDelay);
+            immutable period_start = Height(period_end - params.BlockRewardGap + 1);
+            Amount remainder;
+
+            // calculate the penalty for missing singatures
+            auto pkey_sigcounts = getPKeysOfSignedNonMissingValidators(period_start, period_end);
+            double total_actual_sigcount = pkey_sigcounts.byValue().fold!((a, b) => a + b)(0);
+            log.trace("Total actual signatures count: {}", total_actual_sigcount);
+            double total_expected_sigcount = iota(period_start.value, period_end.value + 1)
+                .map!(height => this.enroll_man.getCountOfValidators(Height(height))).fold!((a, b) => a + b)(cast(size_t) 0);
+            log.trace("Total expected signatures count: {}", total_expected_sigcount);
+            immutable comply_perc = (total_actual_sigcount / total_expected_sigcount) * 100;
+
+            // calculate the validator block reward for each validator
+            auto each_validator_reward = this.block_reward_calc.getEachValidatorReward(
+                            height, getPKeysOfSignedNonMissingValidators(period_start, period_end), remainder, comply_perc);
+
+            // calculate the foundation block reward
+            Amount foundation_reward = this.block_reward_calc.getTotalCommonsBudgetReward(height, comply_perc);
+            foundation_reward.mustAdd(remainder);
+            each_validator_reward.update(
+                Point(params.CommonsBudgetAddress[]),
+                {
+                    auto amount = Amount(foundation_reward);
+                    return amount;
+                },
+                (ref Amount amount)
+                {
+                    amount.mustAdd(foundation_reward);
+                    return amount;
+                }
+            );
+
+            // merge the block reward with fee reward
+            foreach (ref pkey_reward; each_validator_reward.byKeyValue())
+                if (pkey_reward.value != Amount(0))
+                {
+                    if (auto pkey_ind_found = (pkey_reward.key in pkey_to_output_ind))
+                        coinbase_tx.outputs[*pkey_ind_found].value.mustAdd(pkey_reward.value);
```

Comment:
`mustAdd` shouldn't be used outside of tests

Reply:
> mustAdd

so if I use 'add', should I check if the resulting Amount is still valid? In case your answer is 'yes I should check' and it is an invalid amount after the addition, then I would like to assert(0). I am confused